homeassistant/components/essent/coordinator.py

```diff
@@ -0,0 +1,310 @@
+            next_hour.year,
+            next_hour.month,
+            next_hour.day,
+            next_hour.hour,
+            tzinfo=dt_util.UTC,
+        )
+
+        _LOGGER.debug("Scheduling next listener tick for %s", next_run)
+
+        @callback
+        def _handle(_: datetime) -> None:
+            """Handle the scheduled listener tick to update sensors."""
+            self._unsub_listener = None
+            _LOGGER.debug("Listener tick fired, updating sensors with cached data")
+            self.async_update_listeners()
+            self._schedule_listener_tick()
+
+        self._unsub_listener = async_track_point_in_utc_time(
+            self.hass,
+            _handle,
+            next_run,
+        )
+
+    def _normalize_energy_block(
+        self,
+        data: dict[str, Any],
+        energy_type: str,
+        tomorrow: dict[str, Any] | None,
+    ) -> EssentEnergyData:
+        """Normalize the energy block into the coordinator format."""
+        tariffs_today = sorted(
+            data.get("tariffs", []),
+            key=_tariff_sort_key,
+        )
+        if not tariffs_today:
+            _LOGGER.debug("No tariffs found for %s in payload: %s", energy_type, data)
+            raise UpdateFailed(f"No tariffs found for {energy_type}")
+
+        tariffs_tomorrow: list[dict[str, Any]] = []
+        if tomorrow:
+            tariffs_tomorrow = sorted(
+                tomorrow.get("tariffs", []),
+                key=_tariff_sort_key,
+            )
+        unit = (data.get("unitOfMeasurement") or data.get("unit") or "").strip()
+
+        amounts = [
+            float(total)
+            for tariff in tariffs_today
+            if (total := tariff.get("totalAmount")) is not None
+        ]
+        if not amounts:
+            _LOGGER.debug(
+                "No usable totalAmount values for %s in tariffs: %s",
+                energy_type,
+                tariffs_today,
+            )
+            raise UpdateFailed(f"No usable tariff values for {energy_type}")
+
+        if not unit:
+            _LOGGER.debug("No unit provided for %s in payload: %s", energy_type, data)
+            raise UpdateFailed(f"No unit provided for {energy_type}")
+
+        return {
+            "tariffs": tariffs_today,
+            "tariffs_tomorrow": tariffs_tomorrow,
+            "unit": _normalize_unit(unit),
+            "min_price": min(amounts),
+            "avg_price": sum(amounts) / len(amounts),
+            "max_price": max(amounts),
+        }
+
+    async def _async_update_data(self) -> EssentData:
+        """Fetch data from API."""
+        session = async_get_clientsession(self.hass)
+        try:
+            response = await session.get(
+                API_ENDPOINT,
+                timeout=CLIENT_TIMEOUT,
+                headers={"Accept": "application/json"},
+            )
+        except ClientError as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
```

Comment:
Service specific logic should be packaged in a library hosted on pypi.

Bonus points if you can look at a library like mashumaro to transform the data you receive in a typesafe dataclass

Reply:
Done - d40b90fe1e0807b3495e8c092a17d1e2b223cbe0